test/test_runner.rb

```diff
@@ -31,57 +34,169 @@ class TestRunner < Minitest::Test
   # The Runner or rather it's storage is stateful so all steps in the test
   # must be made in order to keep the test self contained. Each step is a
   # separate function though.
-  def test_runner_basics
-    http = Net::HTTP.new($host, $port)
+  def test_runner_using_refs
+    client = WAB::Client.new($host, $port)
 
     # Delete all records to start with a clean database
-    clear_records(http)
-    ref = create_record(http)
-    read_record(http, ref)
-    list_records(http, ref)
+    clear_records(client)
 
-    ref = update_record(http, ref)
-    read_after_update(http, ref)
+    ref = create_records(client)
+    get_record(client, ref)
+    update_record(client, ref)
+    delete_record(client, ref)
+  end
+
+  def test_runner_using_queries
+    client = WAB::Client.new($host, $port)
 
-    delete_record(http, ref)
-    read_after_delete(http, ref)
+    clear_records(client)
+    ref = create_records(client)
+
+    read_record(client, ref)
+    list_records(client)
+    query_records(client)
+    update_query(client)
+    delete_query(client)
   end
 
-  def clear_records(http)
-    resp = http.send_request('DELETE', '/v1/Article')
-    # Response should be an OK with a JSON body.
-    assert_equal(Net::HTTPOK, resp.class, 'response not an OK')
-    reply = Oj.strict_load(resp.body, symbol_keys: true)
+  def test_runner_missing
+    client = WAB::Client.new($host, $port)
+
+    clear_records(client)
+    create_records(client)
+
+    puts "  --- read missing" if $VERBOSE
+    result = client.read('Article', {title: 'Missing'})
+    check_result_code(result)
+    assert_equal(0, result[:results].length, 'read results should be empty')
+
+    puts "  --- update missing" if $VERBOSE
+    result = client.update('Article', {kind: 'Article'}, {title: 'Missing'})
+    check_result_code(result)
+    assert_equal(0, result[:updated].length, 'updated should be empty')
+    
+    puts "  --- delete missing" if $VERBOSE
+    result = client.delete('Article', {title: 'Missing'})
+    check_result_code(result)
+    assert_equal(0, result[:deleted].length, 'deleted should be empty')
   end
 
-  # Returns the reference of the created record.
-  def create_record(http)
-    json = %|{
-    "kind": "Article",
-    "title": "Sample",
-    "text": "Just some random text."
-}|
-    resp = http.send_request('PUT', '/v1/Article', json, { 'Content-Type' => 'application/json' })
-    # Response should be an OK with a JSON body.
-    assert_equal(Net::HTTPOK, resp.class, 'response not an OK')
-    reply = Oj.strict_load(resp.body, symbol_keys: true)
+  def test_runner_duplicate
+    client = WAB::Client.new($host, $port)
```

Comment:
Why create a new instance of `WAB;;Client` at each test? Why not create one instance and run multiple tests with it?

Reply:
Initially I was concerned with state maintained in the client but there is none so a single alien would be fine. I'll change to use the `Test.setup()` method but that gets called for every test so it only changes the number of lines of code, not what actually happens. clear_records is also there as well now.